mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp

```diff
@@ -52,6 +52,82 @@ template <typename Ty> static Ty getMemRefElementType(Type ty) {
   return dyn_cast<Ty>(mt.getElementType());
 }
 
+/// Returns true if \p op may have an effect of type \tparam EffectTys on \p
+/// value.
+template <typename... EffectTys>
+static bool mayHaveEffect(Operation &op, Value value) {
+  if (auto memOp = dyn_cast<MemoryEffectOpInterface>(op)) {
+    if (hasEffect<EffectTys...>(&op, value))
+      return true;
+    // If the op does not have recursive side effects, then it does not have
+    // such effects in the value.
+    if (!op.hasTrait<OpTrait::HasRecursiveMemoryEffects>())
+      return false;
+  } else if (!op.hasTrait<OpTrait::HasRecursiveMemoryEffects>()) {
+    // Otherwise, if the op does not implement the memory effect interface and
+    // it does not have recursive side effects, then we safely assume it has the
+    // queried effects on the value.
+    return true;
+  }
+
+  return llvm::any_of(op.getRegions(), [&](Region &region) {
+    return llvm::any_of(region.getOps(), [&](Operation &op) {
+      return mayHaveEffect<EffectTys...>(op, value);
+    });
+  });
+}
+
+/// Test \p id is default constructed and it is not modified.
+static bool isDefaultConstructedID(Value id) {
+  // Check it is default constructed
+  auto constructor = id.getDefiningOp<sycl::SYCLIDConstructorOp>();
+  if (!constructor || !constructor.getArgs().empty())
+    return false;
+
+  // Check it has not been modified after construction
+  return llvm::none_of(id.getUsers(), [=](Operation *op) {
+    return mayHaveEffect<MemoryEffects::Write, MemoryEffects::Free>(*op, id);
+  });
+}
+
+namespace {
+/// Transform:
+/// ```
+/// %off = sycl.id.constructor() : () -> memref<?x!sycl_id_X_>
+/// sycl.nd_range.constructor(%gs, %ls, %off)
+///   : (memref<?x!sycl_range_X_>,
+///      memref<?x!sycl_range_X_>,
+///      memref<?x!sycl_id_X_>)
+///   -> memref<?x!sycl_nd_range_X_>
+/// ```
+/// To:
+/// ```
+/// sycl.nd_range.constructor(%gs, %ls)
+///   : (memref<?x!sycl_range_X_>, memref<?x!sycl_range_X_>)
+///   -> memref<?x!sycl_nd_range_X_>
+/// ```
+class SYCLNDRangeConstructorEraseDefaultOffset
+    : public OpRewritePattern<SYCLNDRangeConstructorOp> {
+public:
+  using OpRewritePattern<SYCLNDRangeConstructorOp>::OpRewritePattern;
+
+  LogicalResult match(SYCLNDRangeConstructorOp op) const final {
+    OperandRange args = op.getArgs();
+    return success(args.size() == 3 &&
```

Comment:
Nit: `offsetOperandIndex` is explicitly defined, should we rely on this definition (`+1`) here?